### PR TITLE
Add sample to extend AuthenticationRequestHandler class

### PR DESCRIPTION
**Pull request by VimukthiPerera**
## Purpose
Add sample to extend AuthenticationRequestHandler

## Goals
Add sample to extend AuthenticationRequestHandler

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

## Related PRs
https://github.com/wso2/samples-is/pull/148

## Test environment
Java 1.7
Ubuntu 18.04

### CODE REVIEW

File: authentication-request-handler/pom.xml [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+
+	<modelVersion>4.0.0</modelVersion>
+	<groupId>org.wso2.carbon.identity.framework.extended</groupId>
+	<artifactId>authentication.request.handler</artifactId>
+	<packaging>jar</packaging>
+	<version>1.0.0</version>
+	<name>sample-request-handler-extension</name>
+	<description></description>
+
+	<!-- TODO
```
Comment (vihanga-liyanage) [inline authentication-request-handler/pom.xml]: Shall we remove these if not needed?

File: authentication-request-handler/pom.xml [hunk 2]
```diff
@@ -0,0 +1,170 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (vihanga-liyanage) [inline authentication-request-handler/pom.xml]: The file is indented with 8 characters, where WSO2 standard is 4. Can please fix it?

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
+package org.wso2.carbon.identity.framework.extended;
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Please add the license header.

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+public class ExtendedAuthenticationRequestHandler extends DefaultAuthenticationRequestHandler {
+
+    private static final String ROLE_CLAIM = "http://wso2.org/claims/role";
+    private static final String ROLE_TO_CHECK = "trip";
+
+    @Override
+    protected void concludeFlow(HttpServletRequest request,
+                                HttpServletResponse response,
+                                AuthenticationContext context) throws FrameworkException {
+
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Please remove the additional new line.

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        boolean isAuthenticated = context.isRequestAuthenticated();
+
+        if (isAuthenticated) {
+            // Do the authorization logic here
+            if (isAuthorized(request, response, context)) {
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: We could just swap the IF statement here.
```
if (!isAuthorized(request, response, context)) {
        context.setRequestAuthenticated(false);
}
```

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    private boolean isAuthorizationEnforcedOnSp(String serviceProviderName) {
+        return StringUtils.equalsIgnoreCase(serviceProviderName, "TripAction");
+    }
+
+    private boolean isSamlLogin(AuthenticationContext context) {
+
+        return StringUtils.equalsIgnoreCase(context.getRequestType(),
+                FrameworkConstants.RequestType.CLAIM_TYPE_SAML_SSO);
+    }
+}
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Add a new line at EOF.

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            return new ArrayList<>();
+        } else {
+            String roleClaim = userAttributes.get(roleClaimMapping);
+            String[] roles = roleClaim.split(attributeSeparator);
+            return new ArrayList<>(Arrays.asList(roles));
+        }
+    }
+
+    private boolean isAuthorizationEnforcedOnSp(String serviceProviderName) {
+        return StringUtils.equalsIgnoreCase(serviceProviderName, "TripAction");
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Shall we make the "TripAction" a constant?

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                // Check whether user has the required role
+                return userRoles.contains(ROLE_TO_CHECK);
+            }
+        }
+
+        return true;
+    }
+
+    private List<String> getUserRoles(AuthenticatedUser authenticatedUser) {
+        Map<ClaimMapping, String> userAttributes = authenticatedUser.getUserAttributes();
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Add a new line at method body start.

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        ClaimMapping roleClaimMapping = null;
+        String attributeSeparator = ",";
+        for (ClaimMapping claimMapping : userAttributes.keySet()) {
+            if (StringUtils.equals(claimMapping.getLocalClaim().getClaimUri(), ROLE_CLAIM)) {
+                roleClaimMapping = claimMapping;
+                break;
+            }
+            if(claimMapping.getLocalClaim().getClaimUri().equals("MultiAttributeSeparator")) {
+                attributeSeparator = userAttributes.get(claimMapping);
+            }
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: According to this logic, is the ROLE_CLAIM appears before the MultiAttributeSeparator separator claim, the code will exit from the for loop and attributeSeparator will not get changed. Is this the intended behavior?

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+            // Role claim is not in the user attributes
+            return new ArrayList<>();
+        } else {
+            String roleClaim = userAttributes.get(roleClaimMapping);
+            String[] roles = roleClaim.split(attributeSeparator);
+            return new ArrayList<>(Arrays.asList(roles));
+        }
+    }
+
+    private boolean isAuthorizationEnforcedOnSp(String serviceProviderName) {
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Add a new line at method body start.

File: authentication-request-handler/src/main/resources/log4j.properties [hunk 11]
```diff
@@ -0,0 +1,12 @@
+log4j.appender.Stdout=org.apache.log4j.ConsoleAppender
```
Comment (vihanga-liyanage) [inline authentication-request-handler/src/main/resources/log4j.properties]: Add license.

Why do we have to have a separate log4j properties file here? Can't we use the main one?

File: authentication-request-handler/pom.xml [hunk 12]
```diff
@@ -0,0 +1,170 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Comment (vihanga-liyanage) [inline authentication-request-handler/pom.xml]: Please add the license header.

Comment (VimukthiPerera) [reply authentication-request-handler/pom.xml]: Fixed with c907fcc2a908e093c988f8bf86c54ce8e98b72c0

File: authentication-request-handler/src/main/resources/log4j.properties (hunk 11 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/resources/log4j.properties]: Fixed with c907fcc2a908e093c988f8bf86c54ce8e98b72c0

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java (hunk 10 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Fixed with c907fcc2a908e093c988f8bf86c54ce8e98b72c0

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java (hunk 8 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Fixed with c907fcc2a908e093c988f8bf86c54ce8e98b72c0

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java (hunk 6 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Fixed with c907fcc2a908e093c988f8bf86c54ce8e98b72c0

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java (hunk 3 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Fixed with c907fcc2a908e093c988f8bf86c54ce8e98b72c0

File: authentication-request-handler/pom.xml (hunk 1 above)
Comment (VimukthiPerera) [reply authentication-request-handler/pom.xml]: Fixed with c907fcc2a908e093c988f8bf86c54ce8e98b72c0

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java (hunk 5 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Fixed with d3c4ba88995f4b268756fa96d439f275f112d370

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java (hunk 4 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Fixed with d3c4ba88995f4b268756fa96d439f275f112d370

File: authentication-request-handler/pom.xml (hunk 2 above)
Comment (VimukthiPerera) [reply authentication-request-handler/pom.xml]: Fixed with ef2449582ec210c80807bb2f728fec586f3846f2

File: authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java (hunk 9 above)
Comment (VimukthiPerera) [reply authentication-request-handler/src/main/java/org/wso2/carbon/identity/framework/extended/ExtendedAuthenticationRequestHandler.java]: Fixed with ef2449582ec210c80807bb2f728fec586f3846f2

**Pull request merged**